webdriver/tests/execute_async_script/collections.py

```diff
@@ -157,3 +157,34 @@ def test_node_list(session, inline):
     assert len(value) == 2
     for expected, actual in zip(ps, value):
         assert_same_element(session, expected, actual)
+
+
+def test_shadow_root(session, inline):
+    session.url = inline("""
+        <style>
+            custom-checkbox-element {
+                display:block; width:20px; height:20px;
+            }
+        </style>
+        <custom-checkbox-element></custom-checkbox-element>
+        <script>
+            customElements.define('custom-checkbox-element',
+                class extends HTMLElement {
+                    constructor() {
+                            super();
+                            this.attachShadow({mode: 'open'}).innerHTML = `
+                                <div><input type="checkbox"/></div>
+                            `;
+                        }
+                });
+        </script>""")
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    expected = custom_element.shadow_root
+    response = execute_async_script(session, """
+        let resolve = arguments[0];
+        resolve(document.querySelector('custom-checkbox-element').shadowRoot);
+        """)
+    value = assert_success(response)
+    assert isinstance(value, dict)
+    assert "shadow-6066-11e4-a52e-4f735466cecf" in value
+    assert value["shadow-6066-11e4-a52e-4f735466cecf"] == expected.id
```

Comment:
Shouldn't we be able to use `assert_same_element(session, expected, actual)` here?

Reply:
Alas, no, because `assert_same_element` evaluates either that the passed in values are instances of `Element` (which they are not, because they are shadow roots), or, if they are a `dict`, that they contain the element identifier of `element-6066-11e4-a52e-4f735466cecf` (which they do not, also because they are shadow roots).

Reply:
Ah, I see. As such best might be to have a new `assert_same_shadow_root()` helper then. But it's not something that we would have to do now. Feel free to add if you wanna see it correctly covered. Other shadow root tests would also benefit from it.